---
title: "Leiomyoma_RNAseq_GWAS_scRNAseq"
author: "Alexander J. Duval"
date: "2023-07-17"
output= "github_document"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Load environment
```{r}
silence <- suppressPackageStartupMessages
silence(library(ggplot2))
silence(library(ggrepel))
silence(library(grid))
silence(library(dplyr))
silence(library(DESeq2))
silence(library(pheatmap))
silence(library(biomaRt))
silence(library(ggVennDiagram))
silence(library(UpSetR))
silence(library(Seurat))
```

## GSE169255 RNA-seq analysis
```{r}
# load count matrix
gse169255 <- read.table("GSE169255_sample_id_ReadsperGene.txt", sep = "\t", header = T)

row.names(gse169255) <- gse169255$gene_id

gse169255 <- gse169255[, -1]

# remove unpaired samples
gse169255 <- gse169255[, -c(1:6)]

# create coldata
gse169255_coldata <- data.frame(cbind(c(rep("myometrium", 12), rep("fibroid", 6)), c(rep("unpaired", 6), rep("paired", 12))))
gse169255_coldata <- filter(gse169255_coldata, X2 != "unpaired")
row.names(gse169255_coldata) <- colnames(gse169255)
colnames(gse169255_coldata) <- c("tissue", "paired_status")

# differential expression analysis
dds_gse169255 <- DESeqDataSetFromMatrix(countData = gse169255,
                              colData = gse169255_coldata,
                              design = ~ tissue)


keep_gse169255 <- rowSums(counts(dds_gse169255)) >= 10
dds_gse169255 <- dds_gse169255[keep_gse169255,]


dds_gse169255$tissue <- factor(dds_gse169255$tissue, levels = c("myometrium","fibroid"))

dds_gse169255 <- DESeq(dds_gse169255)
res_gse169255 <- results(dds_gse169255, pAdjustMethod = 'fdr')
vsd_gse169255 <- vst(dds_gse169255, blind = FALSE)

res_gse169255_df <- data.frame(res_gse169255)


# pick DEGs
gse169255_sig_up <- row.names(filter(res_gse169255_df, log2FoldChange > 0 & padj < 0.05))

gse169255_sig_down <- row.names(filter(res_gse169255_df, log2FoldChange < 0 & padj < 0.05))

gse169255_deg <- c(gse169255_sig_up, gse169255_sig_down)

myColor <- colorRampPalette(c('blue','white', 'red'))(100)

pheatmap(assay(vsd_gse169255)[gse169255_deg,],
                    cluster_rows = TRUE, 
                    show_rownames = FALSE,
                    cluster_cols = TRUE, 
                    annotation_col = gse169255_coldata, 
                    scale = 'row',
                    border_color = NA,
                    color = myColor)
```

## Analyze Debu's RNA-seq data
```{r}
# load data
debu_rna_counts <- read.table("Debu_rna_seq_leio_myo_counts.txt", header = T)

row.names(debu_rna_counts) <- debu_rna_counts[, 1]
debu_rna_counts <- debu_rna_counts[, -1, FALSE]

# create coldata
debu_coldata <- data.frame(c(rep("myometrium", 15), rep("fibroid", 15)))
row.names(debu_coldata) <- colnames(debu_rna_counts)
colnames(debu_coldata) <- "tissue"

# DEseq
dds_debu <- DESeqDataSetFromMatrix(countData = debu_rna_counts,
                              colData = debu_coldata,
                              design = ~ tissue)


keep <- rowSums(counts(dds_debu)) >= 10
dds_debu <- dds_debu[keep,]


dds_debu$tissue <- factor(dds_debu$tissue, levels = c("myometrium","fibroid"))

dds_debu <- DESeq(dds_debu)
res_debu <- results(dds_debu, pAdjustMethod = 'fdr')
vsd_debu <- vst(dds_debu, blind = FALSE)

res_debu_df <- data.frame(res_debu)


# pick DEGs
debu_sig_up <- row.names(filter(res_debu_df, log2FoldChange > 0 & padj < 0.05))

debu_sig_down <- row.names(filter(res_debu_df, log2FoldChange < 0 & padj < 0.05))

debu_deg <- c(debu_sig_up, debu_sig_down)

pheatmap(assay(vsd_debu)[debu_deg,],
                    cluster_rows = TRUE, 
                    show_rownames = FALSE,
                    cluster_cols = TRUE, 
                    annotation_col = debu_coldata, 
                    scale = 'row',
                    border_color = NA,
                    color = myColor)

```

### Get gene symbol annotation data
```{r}
#Get gene symbols from ensembl IDs
ensembl <- useMart("ensembl")
datasets <- listDatasets(ensembl)
ensembl <- useDataset("hsapiens_gene_ensembl", mart=ensembl)


geneNames <- getBM(attributes=c('ensembl_gene_id', 'external_gene_name'), 
      values = unique(c(row.names(res_gse169255_df), row.names(res_debu_df))),
      mart = ensembl)
```

### Functions to query ENSEMBL IDs and return Official Gene Symbols and vice versa
```{r}
GetGeneNames <- function(ensemblID){
  ifelse(ensemblID%in%geneNames$ensembl_gene_id,
         return(geneNames[which(geneNames==ensemblID, arr.ind = TRUE)[,1],2]),
         return(ensemblID))
}

GetEnsembl <- function(geneSymbol){
  ifelse(geneSymbol%in%geneNames$external_gene_name,
         return(geneNames[which(geneNames$external_gene_name == geneSymbol), ]$ensembl_gene_id),
         retrun(geneSymbol))
}
```


## Read in GWAS candidate gene data
```{r}
gwas <- read.table("Kadir_FUMA_job89348_050823/genes.txt", header = T)

gwas_genes <- unique(gwas$symbol)
gwas_ensembl <- unique(gwas$ensg)

```



## Find commonly up and down genes across RNA datasets and compare to GWAS genes
```{r}
## all DEGs between both studies
all_degs <- unique(c(gse169255_sig_up, gse169255_sig_down, debu_sig_up, debu_sig_down))

## common genes between all DEGs and top GWAS genes (of 394)
all_degs_gwas <- unique(gwas[gwas$ensg %in% all_degs,]$symbol)

all_degs_gwas_ensg <- unique(gwas[gwas$ensg %in% all_degs,]$ensg)

## all up DEGs between both studies
all_up_degs <- unique(c(gse169255_sig_up, debu_sig_up))

## common genes between all up DEGs and top GWAS genes
all_up_degs_gwas <- unique(gwas[gwas$ensg %in% all_up_degs, ]$symbol)


## all down DEGs between both studies
all_down_degs <- unique(c(gse169255_sig_down, debu_sig_down))

## common genes between all up DEGs and top GWAS genes
all_down_degs_gwas <- unique(gwas[gwas$ensg %in% all_down_degs, ]$symbol)

```

## Venn diagram showing overlap of DEGs from both studies as well as GWAS genes
```{r}
ggVennDiagram(list(A = c(debu_sig_up, debu_sig_down), B = c(gse169255_sig_up, gse169255_sig_down), C = gwas$ensg))

```

## Heatmap showing GWAS/RNA-seq common genes and their presence in certain datasets
```{r}
# 0 means not present in data, 1 means present, 2 means upregulated, 3 means downregulated
genomic_position <- ifelse(gwas$posMapSNPs == 0, 0, 1)
eqtl <- ifelse(gwas$eqtlMapSNPs == 0, 0, 1)
chromatin_interaction <- ifelse(gwas$ciMap == "No", 0, 1)

debu_deg <- ifelse(gwas$ensg %in% debu_sig_down, 3, ifelse(gwas$ensg %in% debu_sig_up, 2, 0))
gse169255_deg <- ifelse(gwas$ensg %in% gse169255_sig_down, 3, ifelse(gwas$ensg %in% gse169255_sig_up, 2, 0))

gwas_matrix <- cbind(genomic_position, eqtl, chromatin_interaction, debu_deg, gse169255_deg)
gwas_matrix <- t(apply(gwas_matrix, MARGIN = 2, FUN = as.numeric))



colnames(gwas_matrix) <- gwas$symbol
gwas_matrix <- gwas_matrix[,duplicated(colnames(gwas_matrix)) == F]

gwas_matrix <- gwas_matrix[, colnames(gwas_matrix) %in% all_degs_gwas]

gwas_matrix_up <- gwas_matrix[, gwas_matrix[4,] == 2 | gwas_matrix[5,] == 2]

gwas_matrix_down <- gwas_matrix[, gwas_matrix[4,] == 3 | gwas_matrix[5,] == 3]


ComplexHeatmap::Heatmap(gwas_matrix_up, col = c("white", "green4", "red"), cluster_rows = F, rect_gp = gpar(col = "black"), show_heatmap_legend = F)

ComplexHeatmap::Heatmap(gwas_matrix_down, col = c("white", "green4", "blue"), cluster_rows = F, rect_gp = gpar(col = "black"), show_heatmap_legend = F)
```

## Calculate significance of overlap between DEGs and FUMA-identified genes
```{r}
# gse169255 number of candidate DEGS and non-DEGs
gse169255_candidate_DEG <- length(intersect(c(gse169255_sig_up, gse169255_sig_down), gwas$ensg))
gse169255_candidate_nonDEG <- length(gwas$ensg) - gse169255_candidate_DEG

# gse169255 number of non-candidate DEGS and non-DEGS
gse169255_nonCandidate_DEG <- length(c(gse169255_sig_up, gse169255_sig_down)) - gse169255_candidate_DEG
gse169255_nonCandidate_nonDEG <- length(row.names(gse169255)[-c(which(row.names(gse169255) %in% c(gse169255_sig_up, gse169255_sig_down, gwas$ensg)))])

gse169255_fuma_fisher <- data.frame(
  "Candidate" = c(gse169255_candidate_DEG, gse169255_candidate_nonDEG),
  "Non-Candidate" = c(gse169255_nonCandidate_DEG, gse169255_nonCandidate_nonDEG),
  row.names = c("DEG", "Unchanged"),
  stringsAsFactors = F
)

fisher.test(gse169255_fuma_fisher)

# gse169255 number of candidate DEGS and non-DEGs
debu_candidate_DEG <- length(intersect(c(debu_sig_up, debu_sig_down), gwas$ensg))
debu_candidate_nonDEG <- length(gwas$ensg) - debu_candidate_DEG

# debu number of non-candidate DEGS and non-DEGS
debu_nonCandidate_DEG <- length(c(debu_sig_up, debu_sig_down)) - debu_candidate_DEG
debu_nonCandidate_nonDEG <- length(row.names(debu_rna_counts)[-c(which(row.names(debu_rna_counts) %in% c(debu_sig_up, debu_sig_down, gwas$ensg)))])

debu_fuma_fisher <- data.frame(
  "Candidate" = c(debu_candidate_DEG, debu_candidate_nonDEG),
  "Non-Candidate" = c(debu_nonCandidate_DEG, debu_nonCandidate_nonDEG),
  row.names = c("DEG", "Unchanged"),
  stringsAsFactors = F
)

fisher.test(debu_fuma_fisher)

```

## Compare Gallagher, UK Biobank, and Japan Biobank studies
```{r}
# look at risk loci overlaps
gallagher_risk_loci <- read.table("Kadir_FUMA_job89348_050823/GenomicRiskLoci.txt", header = T)
gallagher_risk_loci$dataset <- rep("gallagher", n = nrow(gallagher_risk_loci))

japan_biobank_risk_loci <- read.table("Japan_Biobank/FUMA_job259732/GenomicRiskLoci.txt", header = T)
japan_biobank_risk_loci$dataset <- rep("japan", n = nrow(japan_biobank_risk_loci))

uk_biobank_risk_loci <- read.table("UKB/ukbio_round2_FUMA_job253818/GenomicRiskLoci.txt", header = T)
uk_biobank_risk_loci$dataset <- rep("uk", n = nrow(uk_biobank_risk_loci))

# the variable window is defined as half the size that we want
window <- 10000

# japan/gallagher
japan_gallagher_overlap_risk_loci <- data.frame()
for (i in 1:23){
  chr <- i
  japan_data <- japan_biobank_risk_loci[japan_biobank_risk_loci$chr == chr,]
  gallagher_data <- gallagher_risk_loci[gallagher_risk_loci$chr == chr,]
  if (is.null(dim(japan_data)) | is.null(dim(gallagher_data))){
    next()
  }
  for (j in 1:nrow(japan_data)){
    locus <- japan_data[j,]
    locus_start <- locus$start
    locus_end <- locus$end
    start_abs <- abs(locus_start - gallagher_data$start)
    end_abs <- abs(locus_end - gallagher_data$end)
    gallagher_overlap_start <- gallagher_data[which(start_abs < window | end_abs < window),]$start
    gallagher_overlap_end <- gallagher_data[which(start_abs < window | end_abs < window),]$end
    if (length(gallagher_overlap_start) == 0 | length(gallagher_overlap_end) == 0){
      next()
    }
    japan_gallagher_overlap_risk_loci <- rbind(japan_gallagher_overlap_risk_loci, c(paste0("chr", chr), locus$start, locus$end, gallagher_overlap_start, gallagher_overlap_end))
  }
}

colnames(japan_gallagher_overlap_risk_loci) <- c("chr", "japan_biobank_start", "japan_biobank_end", "gallagher_start", "gallagher_end")

write.table(japan_gallagher_overlap_risk_loci, file = "japan_biobank_gallagher_gwas_risk_loci_overlap.txt", row.names = F, quote = F)


# uk/gallagher
uk_gallagher_overlap_risk_loci <- data.frame()
for (i in 1:23){
  chr <- i
  uk_data <- uk_biobank_risk_loci[uk_biobank_risk_loci$chr == chr,]
  gallagher_data <- gallagher_risk_loci[gallagher_risk_loci$chr == chr,]
  if (is.null(dim(uk_data)) | is.null(dim(gallagher_data))){
    next()
  }
  for (j in 1:nrow(uk_data)){
    locus <- uk_data[j,]
    locus_start <- locus$start
    locus_end <- locus$end
    start_abs <- abs(locus_start - gallagher_data$start)
    end_abs <- abs(locus_end - gallagher_data$end)
    gallagher_overlap_start <- gallagher_data[which(start_abs < window | end_abs < window),]$start
    gallagher_overlap_end <- gallagher_data[which(start_abs < window | end_abs < window),]$end
    if (length(gallagher_overlap_start) == 0 | length(gallagher_overlap_end) == 0){
      next()
    }
    uk_gallagher_overlap_risk_loci <- rbind(uk_gallagher_overlap_risk_loci, c(paste0("chr", chr), locus$start, locus$end, gallagher_overlap_start, gallagher_overlap_end))
  }
}

colnames(uk_gallagher_overlap_risk_loci) <- c("chr", "uk_biobank_start", "uk_biobank_end", "gallagher_start", "gallagher_end")

write.table(uk_gallagher_overlap_risk_loci, file = "uk_biobank_gallagher_gwas_risk_loci_overlap.txt", row.names = F, quote = F)


#uk/japan
uk_japan_overlap_risk_loci <- data.frame()
for (i in 1:23){
  chr <- i
  uk_data <- uk_biobank_risk_loci[uk_biobank_risk_loci$chr == chr,]
  japan_data <- japan_biobank_risk_loci[japan_biobank_risk_loci$chr == chr,]
  if (is.null(dim(uk_data)) | is.null(dim(japan_data))){
    next()
  }
  for (j in 1:nrow(uk_data)){
    locus <- uk_data[j,]
    locus_start <- locus$start
    locus_end <- locus$end
    start_abs <- abs(locus_start - japan_data$start)
    end_abs <- abs(locus_end - japan_data$end)
    japan_overlap_start <- japan_data[which(start_abs < window | end_abs < window),]$start
    japan_overlap_end <- japan_data[which(start_abs < window | end_abs < window),]$end
    if (length(japan_overlap_start) == 0 | length(japan_overlap_end) == 0){
      next()
    }
    uk_japan_overlap_risk_loci <- rbind(uk_japan_overlap_risk_loci, c(paste0("chr", chr), locus$start, locus$end, japan_overlap_start, japan_overlap_end))
  }
}

colnames(uk_japan_overlap_risk_loci) <- c("chr", "uk_biobank_start", "uk_biobank_end", "japan_start", "japan_end")

write.table(uk_japan_overlap_risk_loci, file = "uk_biobank_japan_biobank_gwas_risk_loci_overlap.txt", row.names = F, quote = F)


# only one genomic loci overlaps all 3 datasets (chr13:40723944), the rest only overlap 2 datasets, remove all overlapping data (making sure not to remove twice)
japan_only_loci <- paste0("japan_only_", seq_along(1:(nrow(japan_biobank_risk_loci) - nrow(japan_gallagher_overlap_risk_loci) - nrow(uk_japan_overlap_risk_loci) + 1)))

gallagher_only_loci <- paste0("gallagher_only_", seq_along(1:(nrow(gallagher_risk_loci) - nrow(japan_gallagher_overlap_risk_loci) - nrow(uk_gallagher_overlap_risk_loci) + 1)))

uk_only_loci <-  paste0("uk_only_", seq_along((1:(nrow(uk_biobank_risk_loci) - nrow(uk_japan_overlap_risk_loci) - nrow(uk_gallagher_overlap_risk_loci) + 1))))

japan_gallagher_loci <- paste0("japan_gallagher_overlap_", seq_along(1:(nrow(japan_gallagher_overlap_risk_loci)-1)))

uk_gallagher_loci <- paste0("uk_gallagher_overlap_", seq_along(1:(nrow(uk_gallagher_overlap_risk_loci)-1)))

japan_uk_gallagher_loci <- "uk_gallagher_japan_overlap_1"

all_genomic_risk_loci_vector <- c(japan_only_loci, gallagher_only_loci, uk_only_loci, japan_gallagher_loci, uk_gallagher_loci, japan_uk_gallagher_loci)

# create binary matrix for upset plot
all_genomic_risk_loci_df <- data.frame(matrix(ncol = 3, nrow = length(all_genomic_risk_loci_vector)))
for (i in 1:length(all_genomic_risk_loci_vector)){
  variant <- all_genomic_risk_loci_vector[i]
  data <- ifelse(rep(grepl(pattern = "gallagher_only", x = variant), 3), c(1, 0, 0),
                 ifelse(rep(grepl(pattern = "uk_only", x = variant), 3), c(0, 1, 0),
                        ifelse(rep(grepl(pattern = "japan_only", x = variant), 3), c(0, 0, 1),
                               ifelse(rep(grepl(pattern = "japan_gallagher_overlap", x = variant), 3), c(1, 0, 1),
                                      ifelse(rep(grepl(pattern = "uk_gallagher_overlap", x = variant), 3), c(1, 1, 0),
                                             c(1, 1, 1))))))
  all_genomic_risk_loci_df[i,] <- as.numeric(data)
}

colnames(all_genomic_risk_loci_df) <- c("gallagher", "uk_biobank", "japan_biobank")


# Upset plot!
upset(data.frame(all_genomic_risk_loci_df), sets = c("japan_biobank", "uk_biobank", "gallagher"), number.angles = 30, point.size = 3.5, line.size = 2, 
    mainbar.y.label = "Genomic Risk Loci Intersections", sets.x.label = "Genomic Risk Loci Per Dataset",
    text.scale = c(1.3, 1.3, 1, 1, 2, 0.75), keep.order = T)

# look at candidate SNP overlaps
japan_biobank_candidate_snps <- read.table("Japan_Biobank/FUMA_job259732/snps.txt", header = T)

gallagher_candidate_snps <- read.table("Kadir_FUMA_job89348_050823/snps.txt", header = T)

japan_biobank_gallagher_snp_overlap <- merge(japan_biobank_candidate_snps, gallagher_candidate_snps, by = "rsID")

write.table(japan_biobank_gallagher_snp_overlap, file = "japan_biobank_gallagher_gwas_candidate_snp_overlap.txt", row.names = F, quote = F)
```

## Load single cell RNA-seq data
```{r}
dirs <- list.dirs(path = "GSE162122_RAW")[-1]

data_list <- lapply(dirs, FUN = function(x){
  Read10X(x)
})

names(data_list) <- sapply(dirs, FUN = function(x){
  sub('.*/', '', x)
})

# create seurat object lists by tissue type
seurat_myo_list <- lapply(data_list[grepl(pattern = "Myometrium", names(data_list)) == T], FUN = function(x){
  CreateSeuratObject(counts = x, project = "Myo_Lyo", min.cells = 3, min.features = 200)
})

seurat_fib_list <- lapply(data_list[grepl(pattern = "Fibroid", names(data_list)) == T], FUN = function(x){
  CreateSeuratObject(counts = x, project = "Myo_Lyo", min.cells = 3, min.features = 200)
})


# add patient ids
for (i in seq_along(seurat_myo_list)){
  seurat_myo_list[[i]][["patient.id"]] <- gsub('_.*', '', names(seurat_myo_list[i]))
}

for (i in seq_along(seurat_fib_list)){
  seurat_fib_list[[i]][["patient.id"]] <- gsub('_.*', '', names(seurat_fib_list[i]))
}

```

## Perform QC on scRNA-seq data
```{r}
# calculate percent MT
for (i in seq_along(seurat_myo_list)){
  seurat_myo_list[[i]][["percent.mt"]] <- PercentageFeatureSet(seurat_myo_list[[i]], pattern = "^MT-")
}

for (i in seq_along(seurat_fib_list)){
  seurat_fib_list[[i]][["percent.mt"]] <- PercentageFeatureSet(seurat_fib_list[[i]], pattern = "^MT-")
}

lapply(seurat_myo_list, FUN = function(x){
  VlnPlot(x, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
})

lapply(seurat_fib_list, FUN = function(x){
  VlnPlot(x, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
})

# QC subsetting
lapply(seurat_myo_list, FUN = function(x){
  subset(x, subset = percent.mt < 10)
})

lapply(seurat_fib_list, FUN = function(x){
  subset(x, subset = percent.mt < 10)
})

```

## Normalize data and select integration features
```{r}
seurat_myo_list <- suppressMessages(lapply(seurat_myo_list, function(x){
  SCTransform(x)
}))

seurat_fib_list <- suppressMessages(lapply(seurat_fib_list, function(x){
  SCTransform(x)
}))

# select features that are repeatedly variable across datasets for integration
features_myo <- SelectIntegrationFeatures(object.list = seurat_myo_list)

features_fib <- SelectIntegrationFeatures(object.list = seurat_fib_list)
```

## Perform integration
```{r}
anchors_myo <- suppressMessages(FindIntegrationAnchors(object.list = seurat_myo_list, anchor.features = features_myo))

anchors_fib <- suppressMessages(FindIntegrationAnchors(object.list = seurat_fib_list, anchor.features = features_fib))

# this command creates an 'integrated' data assay
seurat_combined_myo <- suppressMessages(IntegrateData(anchorset = anchors_myo))

seurat_combined_fib <- suppressMessages(IntegrateData(anchorset = anchors_fib))
```

## Analysis of each tissue type seurat object
```{r}
# specify that we will perform downstream analysis on the corrected data note that the
# original unmodified data still resides in the 'RNA' assay
DefaultAssay(seurat_combined_myo) <- "integrated"

DefaultAssay(seurat_combined_fib) <- "integrated"


# Run the standard workflow for visualization and clustering
seurat_combined_myo <- ScaleData(seurat_combined_myo, verbose = FALSE)
seurat_combined_myo <- RunPCA(seurat_combined_myo, npcs = 30, verbose = FALSE)
seurat_combined_myo <- RunUMAP(seurat_combined_myo, reduction = "pca", dims = 1:30)
seurat_combined_myo <- FindNeighbors(seurat_combined_myo, reduction = "pca", dims = 1:30)
seurat_combined_myo <- FindClusters(seurat_combined_myo, resolution = 0.5)

seurat_combined_fib <- ScaleData(seurat_combined_fib, verbose = FALSE)
seurat_combined_fib <- RunPCA(seurat_combined_fib, npcs = 30, verbose = FALSE)
seurat_combined_fib <- RunUMAP(seurat_combined_fib, reduction = "pca", dims = 1:30)
seurat_combined_fib <- FindNeighbors(seurat_combined_fib, reduction = "pca", dims = 1:30)
seurat_combined_fib <- FindClusters(seurat_combined_fib, resolution = 0.5)

# Visualization
DimPlot(seurat_combined_myo, reduction = "umap", label = TRUE, repel = TRUE, group.by = "patient.id") + ggtitle("Normal Myometrium scRNA-seq")

DimPlot(seurat_combined_fib, reduction = "umap", label = TRUE, repel = TRUE, group.by = "patient.id") + ggtitle("Uterine Fibroid scRNA-seq")

```

## Find cluster markers
```{r}
seurat_combined_myo <- PrepSCTFindMarkers(seurat_combined_myo, assay = "SCT")
myo_de_genes <- FindAllMarkers(seurat_combined_myo, assay = "SCT", only.pos = T)


seurat_combined_fib <- PrepSCTFindMarkers(seurat_combined_fib, assay = "SCT")
fib_de_genes <- FindAllMarkers(seurat_combined_fib, assay = "SCT", only.pos = T)

```

## Add module scores for cell type marker genes
```{r}
smc <- c("MYH11", "TAGLN", "ACTA2", "CNN1", "DES", "CALD1", "MYL9", "RGS5", "MYLK")
fibro <- c("VIM", "ALDH1A1", "THY1", "FN1", "DCN", "OGN", "MGP", "COL1A1", "COL1A2", "COL3A1")
endo <- c("PECAM1", "CDH11", "VWF", "CD93", "EGFL7", "ID3", "FLT1", "MCAM")
immune <- c("CD3D", "CD3E", "FCER1G", "MS4A1", "CD79B", "CST7", "GZMB", "FCGR3A", "MS4A7")
epi <- c("EPCAM","MUC1", "CDH1", "CD24", "KRT14", "ANPEP", "CLDN1", "OCLN", "CD24")
cell_cycle <- c("MCM5", "PCNA", "TYMS", "FEN1", "MCM2", "MCM4", "RRM1")
eryth <- c("HBA1", "HBA2", "HBB")

seurat_combined_myo <- AddModuleScore(object = seurat_combined_myo, features = list(smc), name = "smc")
seurat_combined_myo <- AddModuleScore(object = seurat_combined_myo, features = list(fibro), name = "fibro")
seurat_combined_myo <- AddModuleScore(object = seurat_combined_myo, features = list(endo), name = "endo")
seurat_combined_myo <- AddModuleScore(object = seurat_combined_myo, features = list(immune), name = "immune")
seurat_combined_myo <- AddModuleScore(object = seurat_combined_myo, features = list(epi), name = "epi")
seurat_combined_myo <- AddModuleScore(object = seurat_combined_myo, features = list(eryth), name = "eryth")


seurat_combined_fib <- AddModuleScore(object = seurat_combined_fib, features = list(smc), name = "smc")
seurat_combined_fib <- AddModuleScore(object = seurat_combined_fib, features = list(fibro), name = "fibro")
seurat_combined_fib <- AddModuleScore(object = seurat_combined_fib, features = list(endo), name = "endo")
seurat_combined_fib <- AddModuleScore(object = seurat_combined_fib, features = list(immune), name = "immune")
seurat_combined_fib <- AddModuleScore(object = seurat_combined_fib, features = list(epi), name = "epi")
seurat_combined_fib <- AddModuleScore(object = seurat_combined_fib, features = list(cell_cycle), name = "cell_cycle")
seurat_combined_fib <- AddModuleScore(object = seurat_combined_fib, features = list(eryth), name = "eryth")


VlnPlot(seurat_combined_myo, features = c("smc1", "fibro1", "endo1", "immune1", "epi1", "eryth1"))
VlnPlot(seurat_combined_fib, features = c("smc1", "fibro1", "endo1", "immune1", "epi1", "eryth1", "cell_cycle1"))
```

## Identify cell cluster types in individual seurat objects and plot cell frequencies
```{r}
seurat_combined_myo[["cell.type"]] <- sapply(seurat_combined_myo$seurat_clusters, FUN = function(x){
  cell.type <- ifelse(x %in% c(2,11), "smc",
                      ifelse(x %in% c(9,17), "fibroblast",
                             ifelse(x %in% c(0,3,4,5), "endothelial",
                                    ifelse(x %in% c(1,6,7,10,12,13,14,15,16), "immune",
                                           ifelse(x %in% c(8), "epithelial",
                                                  ifelse(x %in% c(18), "erythrocytes", NA))))))
  return(cell.type)
})

seurat_combined_myo[["cell.type"]] <- factor(seurat_combined_myo$cell.type, levels = c("smc", "fibroblast", "epithelial", "immune", "endothelial", "erythrocytes"))

seurat_combined_fib[["cell.type"]] <- sapply(seurat_combined_fib$seurat_clusters, FUN = function(x){
  cell.type <- ifelse(x %in% c(0,1,4,6,16), "smc",
                      ifelse(x %in% c(14), "fibroblast",
                             ifelse(x %in% c(2,7,9,11,12), "endothelial",
                                    ifelse(x %in% c(3, 5, 8, 13, 17), "immune",
                                           ifelse(x == 10, "epithelial",
                                                  ifelse(x %in% c(15), "erythrocytes", "cycling_endothelial"))))))
                                                  
  return(cell.type)
})

seurat_combined_fib[["cell.type"]] <- factor(seurat_combined_fib$cell.type, levels = c("smc", "fibroblast", "epithelial", "immune", "endothelial", "erythrocytes", "cycling_endothelial"))

DimPlot(seurat_combined_fib, group.by = "cell.type")
DimPlot(seurat_combined_myo, group.by = "cell.type")

# make bar charts showing distribution of each cell type
bar_data_myo <- as.data.frame(t(sapply(levels(factor(seurat_combined_myo$cell.type)), FUN = function(x){
  num <- length(seurat_combined_myo$cell.type[which(seurat_combined_myo$cell.type == x)])
  return(c(x,num))
})))

colnames(bar_data_myo) <- c("cell.type", "count")
bar_data_myo$count <- as.numeric(bar_data_myo$count)
bar_data_myo <- mutate(bar_data_myo, percent = (count / sum(bar_data_myo$count)) * 100)
bar_data_myo <- arrange(bar_data_myo, desc(percent))
bar_data_myo$cell.type <- factor(bar_data_myo$cell.type, levels = bar_data_myo$cell.type)

bar_data_fib <- as.data.frame(t(sapply(levels(factor(seurat_combined_fib$cell.type)), FUN = function(x){
  num <- length(seurat_combined_fib$cell.type[which(seurat_combined_fib$cell.type == x)])
  return(c(x,num))
})))

colnames(bar_data_fib) <- c("cell.type", "count")
bar_data_fib$count <- as.numeric(bar_data_fib$count)
bar_data_fib <- mutate(bar_data_fib, percent = (count / sum(bar_data_fib$count)) * 100)
bar_data_fib <- arrange(bar_data_fib, desc(percent))
bar_data_fib$cell.type <- factor(bar_data_fib$cell.type, levels = bar_data_fib$cell.type)


ggplot(bar_data_myo, aes(x = cell.type, y = percent, fill = cell.type)) +
  geom_bar(stat = "identity") +
  scale_x_discrete(guide = guide_axis(angle = 45))

ggplot(bar_data_fib, aes(x = cell.type, y = percent, fill = cell.type)) +
  geom_bar(stat = "identity") +
  scale_x_discrete(guide = guide_axis(angle = 45))


# combine bar data for grouped bar graph
bar_data_myo$tissue <- "myometrium"
bar_data_fib$tissue <- "leiomyoma"

bar_data_all <- rbind(bar_data_myo, bar_data_fib)
# add missing data for myometrium
bar_data_all <- rbind(bar_data_all, c("cycling_endothelial", 0, 0, "myometrium"))

bar_data_all$tissue <- factor(bar_data_all$tissue, levels = c("myometrium", "leiomyoma"))
bar_data_all$percent <- as.numeric(bar_data_all$percent)
bar_data_all$cell.type <- factor(bar_data_all$cell.type, levels = c("smc", "fibroblast", "epithelial", "immune", "endothelial", "erythrocytes", "cycling_endothelial"))

bar_data_all$cell.type <- factor(bar_data_all$cell.type, levels = c("smc", "endothelial", "immune", "epithelial", "fibroblast", "erythrocytes", "cycling_endothelial"))

ggplot(bar_data_all, aes(x = cell.type, y = percent, fill = tissue)) +
  geom_bar(position = "dodge", stat = "identity")
```

## Save seurat objects
```{r}
saveRDS(seurat_combined_myo, file = "GSE162122_scRNA_Myometrium_Combined.RDS")
saveRDS(seurat_combined_fib, file = "GSE162122_scRNA_Fibroid_Combined.RDS")

seurat_combined_myo <- readRDS("GSE162122_scRNA_Myometrium_Combined.RDS")
seurat_combined_fib <- readRDS("GSE162122_scRNA_Fibroid_Combined.RDS")
```

## Plot heatmaps of pseudobulk gene expression of genes identified as differentially expressed between leiomyoma and myometrium
```{r}
# Leiomyoma
fib_up_degs_gwas <- FetchData(seurat_combined_fib, vars = c("cell.type", all_up_degs_gwas))

# create pseudobulk data
fib_up_degs_gwas_pseudo <- data.frame()
for (i in 1:length(levels(factor(fib_up_degs_gwas$cell.type)))){
  cell_type <- levels(factor(fib_up_degs_gwas$cell.type))[i]
  data <- fib_up_degs_gwas[fib_up_degs_gwas$cell.type == cell_type, -1]
  data <- colMeans(data)
  fib_up_degs_gwas_pseudo <- rbind(fib_up_degs_gwas_pseudo, data)
}

row.names(fib_up_degs_gwas_pseudo) <- levels(factor(fib_up_degs_gwas$cell.type))
fib_up_degs_gwas_pseudo <- apply(fib_up_degs_gwas_pseudo, MARGIN = 1, as.numeric)
row.names(fib_up_degs_gwas_pseudo) <- colnames(fib_up_degs_gwas[, -1])

# plot heatmaps
gene_order <- c("QARS", "WRAP53", "GPR162", "PDCD11", "RILPL1", "KDM2B", "CHD3", "DVL2", "SLC24A4", "SNRNP35", "DDX55", "SORCS3", "CACNA1E", "PKP3", "DACT1", "KCNAB3", "MLXIP", "NKD1", "SLC38A4", "FAM101A", "KIF5C", "CNTROB", "KDM1A", "HSPG2", "ELMOD1", "SH3PXD2A", "NEURL4", "PTDSS2", "VMP1", "B4GALNT4", "KANK1", "KNTC1", "USP46", "NPTXR", "ADAM23", "WT1", "SULT1E1", "SEPT4", "MPPED2", "WNT4", "RASL11B", "VCAN", "DEPDC7", "PFAS", "ELOVL4", "HIST1H4H", "HIST1H2BK", "SHMT2", "ERVMER34-1", "DNAH10", "NBPF3", "RAP1GAP", "DNAH2", "TP53", "HIST1H2BD", "TMEM256", "RBX1", "TMEM256-PLSCR3", "TMEM102", "APOBEC3C", "GLIPR2", "F13A1", "C1QA", "C1QC", "APOBEC3F", "GPR160", "SLC38A1", "RGS17", "SALL1", "ATP6V0A2", "MGAT3", "AEN", "CDH2", "LRRC56", "SLC25A15",  "CDCA7", "EPHB2", "TRIP13", "KCNF1", "E2F2", "AURKB", "TOP2A", "PDIA6")

cell_type_order <- c("smc", "fibroblast", "epithelial", "immune", "endothelial", "erythrocytes", "cycling_endothelial")

fib_up_degs_gwas_pseudo <- fib_up_degs_gwas_pseudo[gene_order, cell_type_order]

pheatmap(fib_up_degs_gwas_pseudo[,colnames(fib_up_degs_gwas_pseudo) != "cycling_endothelial"], scale = "row", cluster_rows = F, cluster_cols = F)
pheatmap(fib_down_degs_gwas_pseudo, scale = "row")
```

